src/base_circuitpython/displayio/tile_grid.py

```diff
@@ -0,0 +1,155 @@
+            self.x = position[0]
+            self.y = position[1]
+        else:
+            self.x = x
+            self.y = y
+
+        self.bitmap = bitmap
+        self.pixel_shader = pixel_shader
+        self.default_tile = default_tile
+        self.in_group = False
+
+    # setitem for an index simply gets the index of the bitmap
+    # rather than the tile index
+    def __setitem__(self, index, value):
+        if isinstance(index, tuple):
+            if index[0] >= self.tile_width or index[1] >= self.tile_height:
+                raise IndexError(CONSTANTS.TILE_OUT_OF_BOUNDS)
+
+        self.bitmap[index] = value
+
+    # getitem for an index simply gets the index of the bitmap
+    # rather than the tile index
+    def __getitem__(self, index):
+        if isinstance(index, tuple):
+            if index[0] >= self.tile_width or index[1] >= self.tile_height:
+                raise IndexError(CONSTANTS.TILE_OUT_OF_BOUNDS)
+
+        return self.bitmap[index]
+
+    # methods that are not in the origin class:
+
+    def draw(self, x, y, scale):
+
+        # draw the current bitmap with
+        # appropriate scale on the global bmp_img
+        x = self.x * scale + x
+        y = self.y * scale + y
+
+        new_shape = self.draw_group(
+            x, y, 0, self.tile_height, 0, self.tile_width, scale
+        )
+
+        img.paste(new_shape, (x, y), new_shape)
+
+    def draw_group(self, x, y, y_start, y_end, x_start, x_end, scale):
+        height = y_end - y_start
+        width = x_end - x_start
+
+        this_img = Image.new("RGBA", (width * scale, height * scale), (0, 0, 0, 0))
+        this_img.putalpha(0)
+        this_bmp_img = this_img.load()
+
+        for i in range(y_start, y_end):
+            for j in range(x_start, x_end):
+                x_offset = j * scale
+                y_offset = i * scale
+
+                x_max = min(x_offset + scale, width * scale)
+                y_max = min(y_offset + scale, height * scale)
+
+                curr_val = self.bitmap[j, i]
+                transparent = self.pixel_shader._Palette__contents[curr_val].transparent
+
+                if not transparent and x_offset >= 0 and y_offset >= 0:
+
+                    curr_colour = self.pixel_shader[curr_val]
+                    self.fill_pixel(
+                        curr_val,
+                        curr_colour,
+                        x_offset,
+                        y_offset,
+                        scale,
+                        x_max,
+                        y_max,
+                        this_bmp_img,
+                    )
+
+        return this_img
+
+    # helper method for drawing pixels on bmp_img
+    # given the src, offset, and scale
+    def fill_pixel(
+        self,
+        curr_val,
+        curr_colour,
+        x_offset,
+        y_offset,
+        scale,
+        x_max,
+        y_max,
+        this_bmp_img,
+    ):
+
+        for new_y in range(y_offset, y_max):

```

Comment:
Is this the drawing logic which we are looking to improve?

Reply:
Yup! Pretty much everything following the "draw()" method in TileGrid

Reply:
I see, it looks like we are drawing each pixel every time 😮 (you probably tweaked it by now, I was just curious haha)